python/tank/util/session.py

```diff
@@ -0,0 +1,181 @@
+    """
+    Returns the cached login info if found.
+    :returns: Returns a dictionnary with keys hostname, login and session_token or None
+    """
+    # Retrieve the location of the cached info
+    info_path = _get_cached_login_info_location(base_url)
+    # Nothing was cached, return an empty dictionary.
+    if not os.path.exists(info_path):
+        logger.debug("No cache found at %s", info_path)
+        return {}
+    try:
+        # Read the login information
+        config = SafeConfigParser({"login": None, "session_token": None})
+        config.read(info_path)
+        if not config.has_section("LoginInfo"):
+            logger.debug("No Login info was found")
+            return {}
+
+        login = config.get("LoginInfo", "login", raw=True)
+        session_token = config.get("LoginInfo", "session_token", raw=True)
+
+        if not login or not session_token:
+            logger.debug("Incomplete settings (login:%s, session_token:%s)", login, session_token)
+            return {}
+
+        return {"login": login, "session_token": session_token}
+    except Exception:
+        logger.exception("Exception thrown while loading cached session info.")
+        return {}
+
+
+def _validate_session_token(host, session_token, http_proxy):
+    """
+    Validates the session token by attempting to an authenticated request on the site.
+    :param session_token: Session token to use to connect to the host.
+    :param host: Host for that session
+    :param http_proxy: http_proxy to use to connect. If no proxy is required, provide None or an empty string.
+    """
+    # Connect to the site
+    logger.debug("Creating shotgun instance")
+    sg = Shotgun(
+        host.encode("utf-8"),
+        session_token=session_token.encode("utf-8"),
+        http_proxy=http_proxy.encode("utf-8") if http_proxy else None
+    )
+    try:
+        sg.find_one("HumanUser", [])
+        return sg
+    except AuthenticationFault, e:
+        # Session was expired.
+        logger.exception(e)
+        return None
+
+
+def cache_session_data(host, login, session_token):
+    """
+    Caches the session data for a site and a user.
+    :param host: Site we want to cache a session for.
+    :param login: User we want to cache a session for.
+    :param session_token: Session token we want to cache.
+    """
+    # Retrieve the cached info file location from the host
+    info_path = _get_cached_login_info_location(host)
+    logger.debug("Caching login info at %s...", info_path)
+    # Create a document with the following format:
+    # [LoginInfo]
+    # login=username
+    # session_token=some_unique_id
+    config = SafeConfigParser()
+    config.add_section("LoginInfo")
+    config.set("LoginInfo", "login", login)
+    config.set("LoginInfo", "session_token", session_token)
+    # Write it to disk.
+    with open(info_path, "w") as configfile:
+        config.write(configfile)
+    logger.debug("Cached!")
+
+
+def delete_session_data(host):
```

Comment:
Since there’s no concept of current user at the moment (there is only one user), it wouldn’t make sense to add it to the signature anyway. I’d rather have signature reflect the state of the code than the future state. It makes the code simpler to understand.

From: Manne Öhrström <notifications@github.com<mailto:notifications@github.com>>
Reply-To: shotgunsoftware/tk-core <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Tuesday, February 17, 2015 at 9:35 AM
To: shotgunsoftware/tk-core <tk-core@noreply.github.com<mailto:tk-core@noreply.github.com>>
Cc: JF Boismenu <jean.francois.boismenu@autodesk.com<mailto:jean.francois.boismenu@autodesk.com>>
Subject: Re: [tk-core] For #28030, [security] User/password sign in on tank command with session caching (#129)


In python/tank/util/session.py<https://github.com/shotgunsoftware/tk-core/pull/129#discussion_r24818888>:

> +    logger.debug("Caching login info at %s...", info_path)
> +    # Create a document with the following format:
> +    # [LoginInfo]
> +    # login=username
> +    # session_token=some_unique_id
> +    config = SafeConfigParser()
> +    config.add_section("LoginInfo")
> +    config.set("LoginInfo", "login", login)
> +    config.set("LoginInfo", "session_token", session_token)
> +    # Write it to disk.
> +    with open(info_path, "w") as configfile:
> +        config.write(configfile)
> +    logger.debug("Cached!")
> +
> +
> +def delete_session_data(host):


Once we do per-user sessions, will this take a user parameter as well? Maybe we can add this straight away? Or not?

—
Reply to this email directly or view it on GitHub<https://github.com/shotgunsoftware/tk-core/pull/129/files#r24818888>.